selectors/stats.js

```diff
@@ -0,0 +1,40 @@
+import { createSelector } from 'reselect';
+import { UploadStatus } from '../actions/uploads';
+
+const getLocations = state => state.locations;
+const getUploads = state => state.uploads;
+
+const uploadEntries = createSelector(
+  getUploads,
+  data => Object.entries(data),
+);
+
+export const getFailedLocations = createSelector(
+  [getUploads, uploadEntries, getLocations],
+  (uploads, data, locations) => {
+    const failedLocations = [];
+    data.forEach(([key, v]) => {
+      if (v.status === UploadStatus.failed && locations[key]) {
```

Comment:
🔴 This looks like it would modify the data. It would also appear that you could do a filter and then project the data.

Reply:
Yeah forgot about reference lol. What do you mean by project? I have a solution but might not be as optimal as what you had in mind.

Reply:
```
    const failedLocations = data
      .filter(([key, v]) => v.status === UploadStatus.failed && locations[key])
      .map(([key, v]) => ({
        key;
        location: locations[key];
        error: uploads[key].error
      });
```